Add method naming rule (#23697)
dotnet/docs: docs/fundamentals/code-analysis/style-rules/naming-rules.md

If you don't specify any custom naming rules, the following default styles are used:

- For classes, structs, enumerations, properties, methods, and events with `public`, `private`, `internal`, `protected`, or `protected_internal` accessibility, the default naming style is Pascal case.

> - For interfaces with `public`, `private`, `internal`, `protected`, or `protected_internal` accessibility, the default naming style is Pascal case with a required prefix of **I**.

Reviewer: Should these include `private_protected` as well? @mavasani to confirm.

Author: Yep. I just tested in visual studio

Reviewer 2: Maybe we should just list the accessibilities it doesn't apply to, since the list is getting so long.

Author: @gewarren I don't think that's a good idea actually. I think it's better to be explicit and verbose. If someone is looking for a specific keyword they'll want to see it called out rather than omitted.

Reviewer 2: Actually, these 6 accessibilities [appear to be *all* the accessibilities](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/accessibility-levels). So I'm really not sure why we even need to mention them here. It leaves you thinking, "so what doesn't this apply to?".

Reviewer 2: Maybe if you don't specify any accessibility, the rule doesn't apply? In that case, we could say something like "For methods that have an accessibility level specified, the default naming rule is Pascal case."

Reviewer: > Maybe if you don't specify any accessibility, the rule doesn't apply?

The rule will still apply. For example, a class with no declared accessibility in source will have an effective accessibility of "internal".

Author: That is true, it's all of the accessibilities, but it's not applying to `field` or `variable`. Maybe calling those two out specifically would be useful? @BillWagner

If you don't specify any custom naming rules, the following default styles are used:

> - For classes, structs, enumerations, properties, methods, and events with `public`, `private`, `internal`, `protected`, `private_protected`, or `protected_internal` accessibility, the default naming style is Pascal case.

Reviewer 2: ```suggestion
- For classes, structs, enumerations, properties, methods, and events with any accessibility, the default naming style is Pascal case.
```

Reviewer 3: I like @gewarren 's suggestion. When I read this, you've enumerated *all* the possible access levels. It's simpler to just say that.
 (Historically, I bet that at one time `private` used a different standard).

> - For interfaces with `public`, `private`, `internal`, `protected`, `private_protected`, or `protected_internal` accessibility, the default naming style is Pascal case with a required prefix of **I**.

Reviewer 2: ```suggestion
- For interfaces with any accessibility, the default naming style is Pascal case with a required prefix of **I**.
```
